File: objc/Project4/Project4/ReaderViewController.m
//
//  ReaderViewController.m
//  Project4
//
//  Created by Jinwoo Kim on 2/9/21.
//

#import "ReaderViewController.h"

@implementation ReaderViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    if (self.article == nil) return;
    
    NSURL *url = [NSURL URLWithString:(NSString *)self.article[@"fields"][@"thumbnail"]];
    if (url == nil) return;
    
    [self.imageView loadURL:url];
    self.imageView.layer.borderColor = UIColor.darkGrayColor.CGColor;
    self.imageView.layer.borderWidth = 2;
    self.imageView.layer.cornerRadius = 20;
    
    self.headline.text = self.article[@"fields"][@"headline"];
    self.body.linkTextAttributes = @{NSForegroundColorAttributeName: UIColor.blackColor};
    NSString *articleBody = self.article[@"fields"][@"body"];
    NSString *formattedArticleBody = [self formatHTML:articleBody];
    NSData *articleBodyData = [formattedArticleBody dataUsingEncoding:NSUTF8StringEncoding];
    if (articleBodyData == nil) return;
    
    NSError * _Nullable error;
    NSAttributedString *bodyText = [[NSAttributedString alloc] initWithData:articleBodyData
                                                                    options:@{NSDocumentTypeDocumentAttribute: NSHTMLTextDocumentType}
                                                         documentAttributes:nil
                                                                      error:&error];
    
    if (error) {
        NSLog(@"%@", error.localizedDescription);
        return;
    }
    
    if (bodyText == nil) return;
    self.body.attributedText = bodyText;
    
    //
    
    /*
     UITouchTypeIndirect : 트랙패드, 애플티비 리모컨
     UITouchTypeDirect : 직접 터치
     UITouchTypePencil : 애플펜슬 지원
     */
    self.body.panGestureRecognizer.allowedTouchTypes = @[[NSNumber numberWithUnsignedInteger:UITouchTypeIndirect]];
    [self.body setSelectable:YES];
}

- (NSString *)formatHTML:(NSString *)html {
    NSURL *wrapperURL = [NSBundle.mainBundle URLForResource:@"wrapper" withExtension:@"html"];
    if (wrapperURL == nil) return html;
    
    NSError * _Nullable error;
    NSString *wrapper = [NSString stringWithContentsOfURL:wrapperURL encoding:NSUTF8StringEncoding error:nil];
    if (error) {
        NSLog(@"%@", error.localizedDescription);
        return html;
    }
    if (wrapper == nil) return html;
    
    return [wrapper stringByReplacingOccurrencesOfString:@"%@" withString:html];
}

@end

File: swift/Project4/Project4/ReaderViewController.swift
//
//  ReaderViewController.swift
//  Project4
//
//  Created by Jinwoo Kim on 2/8/21.
//

import UIKit

class ReaderViewController: UIViewController {
    @IBOutlet weak var headline: UILabel!
    @IBOutlet weak var imageView: RemoteImageView!
    @IBOutlet weak var body: UITextView!
    
    var article: JSON?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        guard let article = article else { return }
        
        if let url = article["fields"]["thumbnail"].url {
            imageView.load(url)
            imageView.layer.borderColor = UIColor.darkGray.cgColor
            imageView.layer.borderWidth = 2
            imageView.layer.cornerRadius = 20
        }
        
        headline.text = article["fields"]["headline"].stringValue
        body.linkTextAttributes = [NSAttributedString.Key.foregroundColor: UIColor.black]
        let articleBody = article["fields"]["body"].stringValue
        let formattedArticleBody = formatHTML(articleBody)
        
        if let articleBodyData = formattedArticleBody.data(using: .utf8) {
            if let bodyText = try? NSAttributedString(data: articleBodyData, options: [NSAttributedString.DocumentReadingOptionKey.documentType: NSAttributedString.DocumentType.html], documentAttributes: nil) {
                body.attributedText = bodyText
            }
        }
        
        //
        
        body.panGestureRecognizer.allowedTouchTypes = [UITouch.TouchType.indirect.rawValue] as [NSNumber]
        body.isSelectable = true
    }
    
    func formatHTML(_ html: String) -> String {
        guard let wrapperURL = Bundle.main.url(forResource: "wrapper", withExtension: "html") else {
            return html
        }
        guard let wrapper = try? String(contentsOf: wrapperURL) else {
            return html
        }
        return wrapper.replacingOccurrences(of: "%@", with: html)
    }
}
